Applications/Utils/ModelPreparation/PartitionMesh/NodeWiseMeshPartitioner.h

```diff
@@ -0,0 +1,197 @@
+#define NODE_WISE_MESH_PARTITIONER_H_
+
+#include <memory>
+#include <tuple>
+#include <vector>
+#include <string>
+#include <fstream>
+
+#include "MeshLib/Mesh.h"
+
+namespace ApplicationUtils
+{
+///  A subdomain mesh.
+struct Partition
+{
+    std::vector<MeshLib::Node*> nodes;  ///< nodes.
+    std::size_t number_of_non_ghost_base_nodes;
+    std::size_t number_of_non_ghost_nodes;
+    std::size_t number_of_base_nodes;
+    /// Non ghost elements
+    std::vector<const MeshLib::Element*> regular_elements;
+    std::vector<const MeshLib::Element*> ghost_elements;
+};
+
+/// Mesh partitioner.
+class NodeWiseMeshPartitioner
+{
+public:
+    typedef long IntegerType;
+
+public:
+    /*!
+     * \param file_name_base The prefix of the file name.
+     * \param num_partitions Number of partitions,
+     * \param mesh           Pointer to a mesh object.
+     */
+    NodeWiseMeshPartitioner(const IntegerType num_partitions,
+                            std::unique_ptr<MeshLib::Mesh>& mesh)
+        : _npartitions(num_partitions),
+          _partitions(num_partitions),
+          _mesh(std::move(mesh)),
+          _nodes_global_ids(_mesh->getNumberOfNodes()),
+          _nodes_partition_ids(_mesh->getNumberOfNodes()),
+          _elements_status(_mesh->getNumberOfElements(), false)
+    {
+    }
+    ~NodeWiseMeshPartitioner() = default;
+
+    /// Partition by node.
+    /// \param is_mixed_hl_elem Flag to indicate whether the elements of
+    /// a mesh can be used for both linear and high order interpolation
+    void partitionByMETIS(const bool is_mixed_hl_elem);
+
+    void resetGlobalNodeIndices();
+
+    /// Read metis data
+    /// \param file_name_base The prefix of the file name.
+    void readMetisData(const std::string& file_name_base);
+
+    /// Write mesh to METIS input file
+    /// \param file_name File name with an extension of mesh.
+    void writeMETIS(const std::string& file_name);
+
+    /// Write the partitions into ASCII files
+    /// \param file_name_base The prefix of the file name.
+    void writeASCII(const std::string& file_name_base);
+
+    /// Write the partitions into binary files
+    /// \param file_name_base The prefix of the file name.
+    void writeBinary(const std::string& file_name_base);
+
+    /// Write the global mesh into a VTU file
+    /// \param file_name_base The prefix of the file name.
+    void writeGlobalMeshVTU(const std::string& file_name_base);
+
+private:
+    /// Number of partitions.
+    IntegerType _npartitions;
+
+    /// Data for all  partitions.
+    std::vector<Partition> _partitions;
+
+    /// Pointer to a mesh object.
+    std::unique_ptr<MeshLib::Mesh> _mesh;
+
+    /// Global IDs of all nodes after partitioning.
+    std::vector<std::size_t> _nodes_global_ids;
+
+    /// Partition IDs of each nodes.
+    std::vector<std::size_t> _nodes_partition_ids;
+
+    /// Flags to indicate the status of all elements.
+    std::vector<bool> _elements_status;
```

Comment:
Looks like "status" means if the element has already been processed. Or do I misunderstand it?
Maybe change comment:
'/// Flags to indicate which elements have already been processed entirely.`